File: core/src/vct/col/ast/DeclarationStatement.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import java.util.*;

import vct.util.ClassName;

public class DeclarationStatement extends ASTDeclaration {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  // Type should become part of ASTNode!!
  private Type type;
  private ASTNode expr;

  public DeclarationStatement(String name,Type type){
    super(name);
    this.type=type;
    expr=null;
  }

  public DeclarationStatement(String name,Type type,ASTNode expr){
    super(name);
    this.type=type;
    this.expr=expr;
  }
  
  public Type getType() { return type; }

  public String getName() { return name; }
  
  public ASTNode getInit() { return expr; }
  
  @Override
  public ClassName getDeclName() {
    hre.System.Debug("%s.%s",package_name,name);
    return new ClassName(package_name,name);
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
      }
      throw t;
    }
  }
 

}

File: vercors/src/main/java/vct/col/ast/stmt/decl/DeclarationStatement.scala
package vct.col.ast.stmt.decl

import hre.lang.System.Debug
import vct.col.ast.`type`.Type
import vct.col.ast.generic.ASTNode
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}
import vct.util.ClassName

/**
 * AST node that represents a declaration statement, for example "`int test := 2+4;`".
 *
 * @author sccblom, whmoortwijn
 * @param name The name of the declared variable, e.g. "`test`".
 * @param type The type of the declared variable, e.g. "`int`".
 * @param init Optionally, an expression that determines the initial value of the declared variable, e.g. "`2+4`".
 */
case class DeclarationStatement(override val name:String, val `type`:Type, val init:Option[ASTNode]) extends ASTDeclaration(name) with VisitorHelper {
  /**
   * Initialises a new AST node that represents a declaration statement without
   * initial value, for example "`int test;`".
   */
  def this(name:String, t:Type) = this(name, t, None)

  /** Builds a new declaration from an initial expression (`init`) that is possibly `null`. */
  def this(name:String, t:Type, init:ASTNode) = this(name, t, Option(init))

  /** Yields the declaration type */
  override def getType = `type`

  /** Java wrapper over the initial expression, which may return `null` (in case `init` equals `None`).  */
  def initJava = init match {
    case Some(node) => node
    case None => null
  }

  /** Yields the full name of the declared variable, including package name. */
  override def getDeclName = {
    Debug(s"$packageName.$name")
    new ClassName(packageName, name)
  }

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields(): Iterable[String] = Seq("type", "init")
  override def debugTreePropertyFields(): Iterable[String] = Seq("name")
}
